Make custom trait object for `Future` generic

MajorBreakfast opened the pull request:
- `TaskObj` -> `FutureObj<'static, ()>`
- The `impl From<...> for FutureObj<'a, T>` impls are impossible because of the type parameter `T`. The impl has to live in libstd, but `FutureObj<'a, T>` is from libcore. Therefore `Into<FutureObj<'a, T>>` was implemented instead. Edit: This didn‘t compile without warnings. I am now using non-generic Form impls.

See https://github.com/rust-lang-nursery/futures-rs/issues/1058

r? @cramertj 

Edit: Added lifetime

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/398608514) ([raw log](https://api.travis-ci.org/v3/job/398608514/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:03:23]    Compiling rustc_tsan v0.0.0 (file:///checkout/src/librustc_tsan)
[00:03:36]    Compiling libc v0.0.0 (file:///checkout/src/rustc/libc_shim)
[00:03:36]    Compiling alloc v0.0.0 (file:///checkout/src/liballoc)
[00:03:36]    Compiling std_unicode v0.0.0 (file:///checkout/src/libstd_unicode)
[00:03:37] warning: conflicting implementations of trait `core::convert::Into<core::task::FutureObj<_>>` for type `boxed::PinBox<_>`: (E0119)
[00:03:37]    --> liballoc/boxed.rs:953:1
[00:03:37]     |
[00:03:37] 953 | impl<T, F: Future<Output = T> + Send + 'static> Into<FutureObj<T>> for PinBox<F> {
[00:03:37]     |
[00:03:37]     = note: #[warn(incoherent_fundamental_impls)] on by default
[00:03:37]     = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
[00:03:37]     = note: for more information, see issue #46205 <https://github.com/rust-lang/rust/issues/46205>
[00:03:37]     = note: for more information, see issue #46205 <https://github.com/rust-lang/rust/issues/46205>
[00:03:37]     = note: conflicting implementation in crate `core`:
[00:03:37]             - impl<T, U> core::convert::Into<U> for T
[00:03:37]               where U: core::convert::From<T>;
[00:03:37]     = note: downstream crates may implement trait `core::convert::From<boxed::PinBox<_>>` for type `core::task::FutureObj<_>`
[00:03:37] 
[00:03:37] warning: conflicting implementations of trait `core::convert::Into<core::task::FutureObj<_>>` for type `boxed::Box<_>`: (E0119)
[00:03:37]    --> liballoc/boxed.rs:960:1
[00:03:37]     |
[00:03:37] 960 | impl<T, F: Future<Output = T> + Send + 'static> Into<FutureObj<T>> for Box<F> {
[00:03:37]     |
[00:03:37]     = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
[00:03:37]     = note: for more information, see issue #46205 <https://github.com/rust-lang/rust/issues/46205>
[00:03:37]     = note: conflicting implementation in crate `core`:
[00:03:37]     = note: conflicting implementation in crate `core`:
[00:03:37]             - impl<T, U> core::convert::Into<U> for T
[00:03:37]               where U: core::convert::From<T>;
[00:03:37]     = note: downstream crates may implement trait `core::convert::From<boxed::Box<_>>` for type `core::task::FutureObj<_>`
[00:03:37] 
[00:03:37] warning: conflicting implementations of trait `core::convert::Into<core::task::LocalFutureObj<_>>` for type `boxed::PinBox<_>`: (E0119)
[00:03:37]    --> liballoc/boxed.rs:967:1
[00:03:37]     |
[00:03:37] 967 | impl<T, F: Future<Output = T> + 'static> Into<LocalFutureObj<T>> for PinBox<F> {
[00:03:37]     |
[00:03:37]     = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
[00:03:37]     = note: for more information, see issue #46205 <https://github.com/rust-lang/rust/issues/46205>
[00:03:37]     = note: conflicting implementation in crate `core`:
[00:03:37]     = note: conflicting implementation in crate `core`:
[00:03:37]             - impl<T, U> core::convert::Into<U> for T
[00:03:37]               where U: core::convert::From<T>;
[00:03:37]     = note: downstream crates may implement trait `core::convert::From<boxed::PinBox<_>>` for type `core::task::LocalFutureObj<_>`
[00:03:37] 
[00:03:37] warning: conflicting implementations of trait `core::convert::Into<core::task::LocalFutureObj<_>>` for type `boxed::Box<_>`: (E0119)
[00:03:37]    --> liballoc/boxed.rs:974:1
[00:03:37]     |
[00:03:37] 974 | impl<T, F: Future<Output = T> + 'static> Into<LocalFutureObj<T>> for Box<F> {
[00:03:37]     |
[00:03:37]     = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
[00:03:37]     = note: for more information, see issue #46205 <https://github.com/rust-lang/rust/issues/46205>
[00:03:37]     = note: conflicting implementation in crate `core`:
[00:03:37]     = note: conflicting implementation in crate `core`:
[00:03:37]             - impl<T, U> core::convert::Into<U> for T
[00:03:37]               where U: core::convert::From<T>;
[00:03:37]     = note: downstream crates may implement trait `core::convert::From<boxed::Box<_>>` for type `core::task::LocalFutureObj<_>`
[00:03:37]    Compiling alloc_system v0.0.0 (file:///checkout/src/liballoc_system)
[00:03:37]    Compiling panic_abort v0.0.0 (file:///checkout/src/libpanic_abort)
[00:03:41]    Compiling panic_unwind v0.0.0 (file:///checkout/src/libpanic_unwind)
[00:03:55]     Finished release [optimized] target(s) in 45.87s
---
[00:21:45]    Compiling rustc_lsan v0.0.0 (file:///checkout/src/librustc_lsan)
[00:22:32]    Compiling libc v0.0.0 (file:///checkout/src/rustc/libc_shim)
[00:22:32]    Compiling alloc v0.0.0 (file:///checkout/src/liballoc)
[00:22:32]    Compiling std_unicode v0.0.0 (file:///checkout/src/libstd_unicode)
[00:22:33] error: conflicting implementations of trait `core::convert::Into<core::task::FutureObj<_>>` for type `boxed::PinBox<_>`: (E0119)
[00:22:33]    --> liballoc/boxed.rs:953:1
[00:22:33]     |
[00:22:33] 953 | impl<T, F: Future<Output = T> + Send + 'static> Into<FutureObj<T>> for PinBox<F> {
[00:22:33]     |
[00:22:33]     |
[00:22:33]     = note: #[deny(incoherent_fundamental_impls)] on by default
[00:22:33]     = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
[00:22:33]     = note: conflicting implementation in crate `core`:
[00:22:33]     = note: conflicting implementation in crate `core`:
[00:22:33]             - impl<T, U> core::convert::Into<U> for T
[00:22:33]               where U: core::convert::From<T>;
[00:22:33]     = note: downstream crates may implement trait `core::convert::From<boxed::PinBox<_>>` for type `core::task::FutureObj<_>`
[00:22:33] 
[00:22:33] error: conflicting implementations of trait `core::convert::Into<core::task::FutureObj<_>>` for type `boxed::Box<_>`: (E0119)
[00:22:33]    --> liballoc/boxed.rs:960:1
[00:22:33]     |
[00:22:33] 960 | impl<T, F: Future<Output = T> + Send + 'static> Into<FutureObj<T>> for Box<F> {
[00:22:33]     |
[00:22:33]     = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
[00:22:33]     = note: for more information, see issue #46205 <https://github.com/rust-lang/rust/issues/46205>
[00:22:33]     = note: conflicting implementation in crate `core`:
[00:22:33]     = note: conflicting implementation in crate `core`:
[00:22:33]             - impl<T, U> core::convert::Into<U> for T
[00:22:33]               where U: core::convert::From<T>;
[00:22:33]     = note: downstream crates may implement trait `core::convert::From<boxed::Box<_>>` for type `core::task::FutureObj<_>`
[00:22:33] 
[00:22:33] error: conflicting implementations of trait `core::convert::Into<core::task::LocalFutureObj<_>>` for type `boxed::PinBox<_>`: (E0119)
[00:22:33]    --> liballoc/boxed.rs:967:1
[00:22:33]     |
[00:22:33] 967 | impl<T, F: Future<Output = T> + 'static> Into<LocalFutureObj<T>> for PinBox<F> {
[00:22:33]     |
[00:22:33]     = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
[00:22:33]     = note: for more information, see issue #46205 <https://github.com/rust-lang/rust/issues/46205>
[00:22:33]     = note: conflicting implementation in crate `core`:
[00:22:33]     = note: conflicting implementation in crate `core`:
[00:22:33]             - impl<T, U> core::convert::Into<U> for T
[00:22:33]               where U: core::convert::From<T>;
[00:22:33]     = note: downstream crates may implement trait `core::convert::From<boxed::PinBox<_>>` for type `core::task::LocalFutureObj<_>`
[00:22:33] 
[00:22:33] error: conflicting implementations of trait `core::convert::Into<core::task::LocalFutureObj<_>>` for type `boxed::Box<_>`: (E0119)
[00:22:33]    --> liballoc/boxed.rs:974:1
[00:22:33]     |
[00:22:33] 974 | impl<T, F: Future<Output = T> + 'static> Into<LocalFutureObj<T>> for Box<F> {
[00:22:33]     |
[00:22:33]     = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
[00:22:33]     = note: for more information, see issue #46205 <https://github.com/rust-lang/rust/issues/46205>
[00:22:33]     = note: conflicting implementation in crate `core`:
[00:22:33]     = note: conflicting implementation in crate `core`:
[00:22:33]             - impl<T, U> core::convert::Into<U> for T
[00:22:33]               where U: core::convert::From<T>;
[00:22:33]     = note: downstream crates may implement trait `core::convert::From<boxed::Box<_>>` for type `core::task::LocalFutureObj<_>`
[00:22:33] error: aborting due to 4 previous errors
[00:22:33] 
[00:22:33] error: Could not compile `alloc`.
[00:22:33] 
[00:22:33] 
[00:22:33] Caused by:
[00:22:33]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name alloc liballoc/lib.rs --color always --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 -C metadata=629be8a8a92e38ff -C extra-filename=-629be8a8a92e38ff --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/release/deps --extern core=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/libcore-e3f2b558276b578b.rlib --extern compiler_builtins=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/libcompiler_builtins-d5d59b383a13b958.rlib -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/build/compiler_builtins-57f7f72e03e89590/out` (exit code: 101)
[00:22:33] error: build failed
[00:22:33] error: build failed
[00:22:33] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "--message-format" "json"
[00:22:33] travis_fold:end:stage1-std


[00:22:33] thread 'main' panicked at 'cargo must succeed', bootstrap/compile.rs:1091:9

[00:22:33] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:22:33] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build
[00:22:33] Build completed unsuccessfully in 0:17:41
[00:22:33] Build completed unsuccessfully in 0:17:41
[00:22:33] make: *** [all] Error 1
[00:22:33] Makefile:28: recipe for target 'all' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:27be0d9e
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:24110906:start=1530374440454319768,finish=1530374440461076786,duration=6757018
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0b1cc28e
$ head -30 ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
head: cannot open ‘./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers’ for reading: No such file or directory
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:0afa7f60
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

ngg commented:
I get the following warnings for the `Into` implementations:
```
warning: conflicting implementations of trait `core::convert::Into<core::task::FutureObj<_>>` for type `boxed::PinBox<_>`: (E0119)
   --> liballoc/boxed.rs:953:1
    |
953 | impl<T, F: Future<Output = T> + Send + 'static> Into<FutureObj<T>> for PinBox<F> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: #[warn(incoherent_fundamental_impls)] on by default
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #46205 <https://github.com/rust-lang/rust/issues/46205>
    = note: conflicting implementation in crate `core`:
            - impl<T, U> core::convert::Into<U> for T
              where U: core::convert::From<T>;
    = note: downstream crates may implement trait `core::convert::From<boxed::PinBox<_>>` for type `core::task::FutureObj<_>`
```
Can't we simply remove that impl?

MajorBreakfast commented:
@nikomatsakis 
@aturon 
@withoutboats 

Can you help? I'm not sure what to do. I can implement neither  `From` nor `Into`

MajorBreakfast commented:
I've removed the `Into` impls again and reverted to `From` impls, but just for `FutureObj<()>` instead of `FutureObj<T>`

<details>
<summary>Into impls that I removed</summary>

```rust
#[unstable(feature = "futures_api", issue = "50547")]
impl<T, F: Future<Output = T> + Send + 'static> Into<FutureObj<T>> for PinBox<F> {
    fn into(self) -> FutureObj<T> {
        FutureObj::new(self)
    }
}

#[unstable(feature = "futures_api", issue = "50547")]
impl<T, F: Future<Output = T> + Send + 'static> Into<FutureObj<T>> for Box<F> {
    fn into(self) -> FutureObj<T> {
        FutureObj::new(PinBox::from(self))
    }
}

#[unstable(feature = "futures_api", issue = "50547")]
impl<T, F: Future<Output = T> + 'static> Into<LocalFutureObj<T>> for PinBox<F> {
    fn into(self) -> LocalFutureObj<T> {
        LocalFutureObj::new(self)
    }
}

#[unstable(feature = "futures_api", issue = "50547")]
impl<T, F: Future<Output = T> + 'static> Into<LocalFutureObj<T>> for Box<F> {
    fn into(self) -> LocalFutureObj<T> {
        LocalFutureObj::new(PinBox::from(self))
    }
}
```
</details>

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/398631990) ([raw log](https://api.travis-ci.org/v3/job/398631990/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Check compiletest suite=run-pass mode=run-pass (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[00:48:13] 
[00:48:13] running 3033 tests
[00:48:27] ....................................................................................................
[00:48:41] ..................F............................i....................................................
[00:49:11] ....................................................................................................
[00:49:24] ....................................................................................................
[00:49:44] ....................................................................................................
[00:49:57] ....................................................................................................
---
[00:57:13] 
[00:57:13] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:498:22
[00:57:13] 
[00:57:13] 
[00:57:13] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/run-pass" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "run-pass" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-3.9/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docc

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:14032df7
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

MajorBreakfast commented:
- Added lifetime
- Moved `FuturesObj` to `core::future`

ngg commented:
I just tried to use it to create async trait methods (including capturing &self with the generic lifetime), it seems to work great.

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/398749360) ([raw log](https://api.travis-ci.org/v3/job/398749360/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:03:27]    Compiling cc v1.0.15
[00:03:27]    Compiling core v0.0.0 (file:///checkout/src/libcore)
[00:03:27]    Compiling build_helper v0.1.0 (file:///checkout/src/build_helper)
[00:03:27]    Compiling unwind v0.0.0 (file:///checkout/src/libunwind)
[00:03:29] error[E0405]: cannot find trait `UnsafeFutureObj` in this scope
[00:03:29]     --> libcore/mem.rs:1232:48
[00:03:29]      |
[00:03:29] 1232 | unsafe impl<'a, T, F: Future<Output = T> + 'a> UnsafeFutureObj<'a, T> for PinMut<'a, F> {
[00:03:29] help: possible candidate is found in another module, you can import it into scope
[00:03:29]      |
[00:03:29]      |
[00:03:29] 18   | use future::future_obj::UnsafeFutureObj;
[00:03:29] 
[00:03:29] 
[00:03:29] error[E0405]: cannot find trait `Future` in this scope
[00:03:29]     --> libcore/mem.rs:1232:23
[00:03:29]      |
[00:03:29] 1232 | unsafe impl<'a, T, F: Future<Output = T> + 'a> UnsafeFutureObj<'a, T> for PinMut<'a, F> {
[00:03:29] help: possible candidate is found in another module, you can import it into scope
[00:03:29]      |
[00:03:29]      |
[00:03:29] 18   | use future::future::Future;
[00:03:29] 
[00:03:29] error[E0412]: cannot find type `Context` in this scope
[00:03:29]     --> libcore/mem.rs:1237:43
[00:03:29]      |
[00:03:29]      |
[00:03:29] 1237 |     unsafe fn poll(ptr: *mut (), cx: &mut Context) -> Poll<T> {
[00:03:29] help: possible candidate is found in another module, you can import it into scope
[00:03:29]      |
[00:03:29] 18   | use task::context::Context;
[00:03:29]      |
[00:03:29]      |
[00:03:29] 
[00:03:29] error[E0412]: cannot find type `Poll` in this scope
[00:03:29]     --> libcore/mem.rs:1237:55
[00:03:29]      |
[00:03:29] 1237 |     unsafe fn poll(ptr: *mut (), cx: &mut Context) -> Poll<T> {
[00:03:29] help: possible candidate is found in another module, you can import it into scope
[00:03:29]      |
[00:03:29] 18   | use task::poll::Poll;
[00:03:29]      |
---
[00:03:30] Caused by:
[00:03:30]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name core libcore/lib.rs --color always --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 -C metadata=c8f7a210a5d40dd6 -C extra-filename=-c8f7a210a5d40dd6 --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/release/deps` (exit code: 101)
[00:03:30] warning: build failed, waiting for other jobs to finish...
[00:03:31] error: build failed
[00:03:31] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "--message-format" "json"
[00:03:31] expected success, got: exit code: 101
[00:03:31] thread 'main' panicked at 'cargo must succeed', bootstrap/compile.rs:1091:9
[00:03:31] travis_fold:end:stage0-std

[00:03:31] travis_time:end:stage0-std:start=1530430443868978546,finish=1530430448871414077,duration=5002435531


[00:03:31] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:03:31] Build completed unsuccessfully in 0:00:06
[00:03:31] make: *** [tidy] Error 1
[00:03:31] Makefile:79: recipe for target 'tidy' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:06dec062
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:12ad5af3:start=1530430449356663655,finish=1530430449363300232,duration=6636577
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:060795bc
$ head -30 ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
head: cannot open ‘./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers’ for reading: No such file or directory
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:002603ab
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

MajorBreakfast commented:
I've added an impl for `PinMut`. It's possible now that `FutureObj` can store a lifetime.

@ngg Thats' great to hear!

ngg commented:
Actually it seems to work properly, it may be sound.

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/398808016) ([raw log](https://api.travis-ci.org/v3/job/398808016/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:03:16]    Compiling rustc_asan v0.0.0 (file:///checkout/src/librustc_asan)
[00:03:19] warning: unused variable: `ptr`
[00:03:19]     --> libcore/mem.rs:1243:20
[00:03:19]      |
[00:03:19] 1243 |     unsafe fn drop(ptr: *mut ()) {}
[00:03:19]      |                    ^^^ help: consider using `_ptr` instead
[00:03:19]      = note: #[warn(unused_variables)] on by default
[00:03:19] 
[00:03:30]    Compiling libc v0.0.0 (file:///checkout/src/rustc/libc_shim)
[00:03:30]    Compiling alloc v0.0.0 (file:///checkout/src/liballoc)
---
[00:21:27]    Compiling rustc_asan v0.0.0 (file:///checkout/src/librustc_asan)
[00:22:00] error: unused variable: `ptr`
[00:22:00]     --> libcore/mem.rs:1243:20
[00:22:00]      |
[00:22:00] 1243 |     unsafe fn drop(ptr: *mut ()) {}
[00:22:00]      |                    ^^^ help: consider using `_ptr` instead
[00:22:00]      = note: `-D unused-variables` implied by `-D warnings`
[00:22:00] 
[00:22:00] 
[00:22:01] Makefile:28: recipe for target 'all' failed
[00:22:01] make: *** [all] Error 1
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

ngg commented:
For example `futures::executor::block_on(obj)` doesn't require `obj` to be `'static`.

MajorBreakfast commented:
@ngg You're right. `FutureObj` with non-static lifetime is definitely useful with `block_on`. I've just tested this and it works with the non-static `FutureObj` I created.

@cramertj 
- Do you think that a `FutureObj` impl for `PinMut` makes sense conceptually?
- I think that `UnsafeFutureObj` should be implemented for `Box` as well. A `PinBox` is IMO not necessary because the future isn't pinned at that point.

MajorBreakfast commented:
@ngg Thanks BTW for all your testing :)

bors commented:
:umbrella: The latest upstream changes (presumably #51969) made this pull request unmergeable. Please resolve the merge conflicts.

cramertj commented on src/libcore/future/future_obj.rs:
```diff
@@ -0,0 +1,150 @@
+    ptr: *mut (),
+    poll_fn: unsafe fn(*mut (), &mut Context) -> Poll<T>,
+    drop_fn: unsafe fn(*mut ()),
+    _marker1: PhantomData<T>,
```
What is this marker for? `T` is already used in the signature of `poll_fn`.

MajorBreakfast replied:
It's for uhm... *kicks it under the rug* ...what marker do you mean? :)

cramertj commented:
@MajorBreakfast Thanks as always for this PR!
> Do you think that a FutureObj impl for PinMut makes sense conceptually?

Yeah, this is great! It's analogous to `PinMut<'a, dyn Future>` or `PinMut<'a, dyn Future + Send>`. I would also add an impl for `&mut T` where `T: Future + Unpin`.

cramertj commented:
r=me with `PhantomType` nit addressed and `&mut` impl added.

mikeyhew commented:
@cramertj they will probably be in nightly within the next month, but not sure when they will be stabilized

cramertj commented:
@mikeyhew Cool! Thanks again for working on that.

@MajorBreakfast I think with that timeline it's still best for us to push forward with this approach and then we can adjust as necessary-- we need both unsized rvalues and object-safe arbitrary self types before we can totally replace these faked-out trait objects.

cramertj commented:
Great!

@bors r+

bors commented:
:pushpin: Commit e666c2bd0742cbf88ff9fa26cfc194099a139589 has been approved by `cramertj`

<!-- @bors r=cramertj e666c2bd0742cbf88ff9fa26cfc194099a139589 -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #51321

bors commented:
:pushpin: Commit e666c2bd0742cbf88ff9fa26cfc194099a139589 has been approved by `cramertj`

<!-- @bors r=cramertj e666c2bd0742cbf88ff9fa26cfc194099a139589 -->

ngg commented:
@cramertj Are you sure that `PhantomData<T>` is really not necessary? This struct does own `T` in the sense that its `Drop` implementation will drop `T`, so I think it's necessary for dropck. (`poll_fn` should be enough for borrowck)

MajorBreakfast commented:
@ngg AFAIK `PhantomData`'s only purpose is to avoid `error[E0392]: parameter T is never used`

MajorBreakfast commented:
@ngg I don't think it is accurate to say that the "struct does own `T`" because `T` is only a type and therefore can't be owned because it's not something that is stored in memory.

ngg commented:
It will not drop `T` when dropped, so my comment was kinda meaningless, sry.
@MajorBreakfast `PhantomData` does lots of other things, it's not just for avoding an error, see https://doc.rust-lang.org/nomicon/phantom-data.html and specifically for dropck uses see https://stackoverflow.com/questions/42708462/why-is-it-useful-to-use-phantomdata-to-inform-the-compiler-that-a-struct-owns-a/42721125#42721125

cramertj commented:
@ngg `PhantomData` isn't needed here-- the type `T` is already used as the output of the `poll_fn`, so that supplies the variance required. The struct itself implements `Drop` manually (always) regardless of the inner type `T`, so adding the `T` here doesn't affect drop checking (AFAIK).

MajorBreakfast commented:
@cramertj I'm not so sure... I think @ngg is right an we're violating this

Here's what `Vec` uses internally:
https://doc.rust-lang.org/1.26.0/src/core/ptr.rs.html#2514-2522

bors commented:
:hourglass: Testing commit e666c2bd0742cbf88ff9fa26cfc194099a139589 with merge c8df60a1462b4b83573573c6e08259a731b60a20...

ngg commented:
Yeah that's what I meant but it wouldn't matter to add `PhantomData<T>` because it's not `T` what we own, but the original Future (let's call it `F`). This whole FutureObj is just emulating `Box<Future<Output = T> + 'a>` which would contain `PhantomData<Future<Output = T> + 'a>` so probably that is what we need here. (I'm really not sure about these things, I'm just guessing.. It might be ok now)

cramertj commented:
@ngg Yeah, if you could spell `PhantomData<Box<dyn UnsafeFutureObj<'a, T>>`, that would be correct, but we don't own a `T`, so `PhantomData<T>` doesn't do anything for us.

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.7997), [status-travis](https://travis-ci.org/rust-lang/rust/builds/399290898?utm_source=github_status&utm_medium=notification)
Approved by: cramertj
Pushing c8df60a1462b4b83573573c6e08259a731b60a20 to master...
